File: nexml/javascript/Phylo/Util/Logger.js
var singleton;

function Logger () {
    if ( singleton == null ) {
        this.level = 2;
        this.fatalLevel = 0;
        this.errorLevel = 1;
        this.warnLevel  = 2;
        this.infoLevel  = 3;
        this.debugLevel = 4;        
        this.listeners  = [ 
            function(msg) {
                self.status = msg;
            }	
        ];
        singleton = this;
	}
	return singleton;
};

Logger.prototype.debug = function(msg) {
	if ( this.level >= this.debugLevel ) {
		this.broadcast( "DEBUG: " + msg );
	}
};

Logger.prototype.info = function(msg) {
	if ( this.level >= this.infoLevel ) {
		this.broadcast( "INFO: " + msg );
	}
};

Logger.prototype.warn = function(msg) {
	if ( this.level >= this.warnLevel ) {
		this.broadcast( "WARN: " + msg );
	}
};

Logger.prototype.error = function(msg) {
	if ( this.level >= this.errorLevel ) {
		this.broadcast( "ERROR: " + msg );
	}
};

Logger.prototype.fatal = function(msg) {
	if ( this.level >= this.fatalLevel ) {
		this.broadcast( "FATAL: " + msg );
	}
};

Logger.prototype.broadcast = function(msg) {
	var listeners = this.listeners;
	for ( var i = 0; i < listeners.length; i++ ) {
		listeners[i](msg);
	}
};

Logger.prototype.VERBOSE = function(level) {
	if ( level >= 0 && level <= 4 ) {
		this.level = level;
	}
};

Logger.prototype.set_listener = function(listener) {
	this.listeners.push(listener);
};

File: java/src/org/biophylo/util/Logger.java
package org.biophylo.util;
import java.util.*;
public class Logger {
	private static Logger mInstance = null;
	private int mLevel;
	private Vector mListeners;
	
	/**
	 * 
	 */
	protected Logger() {
		mLevel = 2;
		mListeners = new Vector();
		mListeners.add(new DefaultListener());
	}
	
	/**
	 * @return
	 */
	public static Logger getInstance() {
		if(mInstance == null) {
			mInstance = new Logger();
		}
	    return mInstance;
	}
	
	/**
	 * @param level
	 */
	public void VERBOSE(int pLevel) {
		mLevel = pLevel;
	}
	
	/**
	 * @param msg
	 */
	public void fatal(String msg) {
		if ( mLevel >= 0 ) {			
			broadcast(msg);
		}
	}
	
	/**
	 * @param msg
	 */
	public void error(String msg) {
		if ( mLevel >= 1 ) {
			broadcast(msg);
		}
	}
	
	/**
	 * @param msg
	 */
	public void warn (String msg) {
		if ( mLevel >= 2 ) {
			broadcast(msg);
		}
	}
	
	/**
	 * @param msg
	 */
	public void info (String msg) {
		if ( mLevel >= 3 ) {
			broadcast(msg);
		}
	}
	
	/**
	 * @param msg
	 */
	public void debug (String msg) {
		if ( mLevel >= 4 ) {
			broadcast(msg);
		}
	}
	
	/**
	 * @param listener
	 */
	public void addListener(LogListener listener) {
		mListeners.add(listener);
	}
	
	/**
	 * @param listener
	 */
	public void removeListener(LogListener listener) {
		mListeners.remove(listener);
	}
	
	/**
	 * @param msg
	 */
	private void broadcast (String msg) {
		Throwable stack = new Throwable();
		StackTraceElement[] stes = stack.getStackTrace();
		StackTraceElement ste = stes[2];
		String[] fullMsg = new String[5];
		fullMsg[0] = msg;
		fullMsg[1] = stes[1].getMethodName().toUpperCase();
		fullMsg[2] = ste.getMethodName();
		fullMsg[3] = ste.getFileName();
		fullMsg[4] = ""+ ste.getLineNumber();
		
		for ( int i = 0; i < mListeners.size(); i++ ) {
			((LogListener)mListeners.get(i)).notify(fullMsg);
		}
	}
}
class DefaultListener implements LogListener {
	int msgI = 0;
	int levelI = 1;
	int methodI = 2;
	int fileI = 3;
	int lineI = 4;
	
	/* (non-Javadoc)
	 * @see org.biophylo.Util.LogListener#notify(java.lang.String[])
	 */
	public void notify (String[] msg) {
		System.err.println(
			msg[levelI]
			    + " " + msg[methodI]
			    + " [" + msg[fileI] + ":" + msg[lineI] + "] - "
			    + msg[msgI]
		);
	}
}
